app/models/git_repository.rb

```diff
@@ -86,6 +92,26 @@ def update_repo
     @updated_repo = true
   end
 
+  # Configures a file lock in LOCKFILE_DIR so that only a single process has
+  # access to make changes to the `GitWorktree` at a time.  Assumes the record
+  # has been saved, since there is no way store (clone, fetch, pull, etc.) the
+  # git data to disk if there isn't a `id`.
+  #
+  # Only a single `@git_lock` can be aquired per-process, and do avoid
+  # deadlocks, his method is just a passthrough if `@git_lock` has already been
+  # defined (another method has already started a `git_transaction`.
+  #
+  # This means that you can surround a couple of actions with this method, and
+  # the lock will only be enforced on the top level.
+  #
+  # Return value is the result of the yielded block
```

Comment:
This all feels like a lot of comment to say "lock the directory to avoid clone collisions".  I don't think we need this entire comment.  locks themselves are pretty clear to most readers.

Reply:
Hey man, I never said that my 📚 were concise, or even grammatically correct (holy typos batman... I was clearly out of it when I wrote this...).

That said, I think at least some things are worth noting based on how I implement it:

The nested transactions bit is important because if it wasn't there you would lock in the outer `git_transaction` and deadlock in the nested one.  I also wanted devs to know that this was designed ensure that the lock was only released in the outer most `git_transaction`, because releasing early would have defeated the purpose.

That said, based on your [other comment](https://github.com/ManageIQ/manageiq/pull/19074#discussion_r309870129), I am more questioning my life choices and less concerned about the length/validity of my comment here...

(in otherwords, this comment being relevant depends first if this implementation is even necessary, and determining if it is concise enough can be answered only after)